core/vm/evm.go

```diff
@@ -502,10 +503,75 @@ func (evm *EVM) Create(caller ContractRef, code []byte, gas uint64, value *big.I
 	if err != nil {
 		return nil, common.ZeroAddr, 0, err
 	}
-	contractAddr = crypto.CreateAddress(caller.Address(), evm.StateDB.GetNonce(internalAddr), code)
+
+	nonce := evm.StateDB.GetNonce(internalAddr)
+
+	contractAddr = crypto.CreateAddress(caller.Address(), nonce, code)
+	if _, err := contractAddr.InternalAddress(); err == nil {
+		return evm.create(caller, &codeAndHash{code: code}, gas, value, contractAddr)
+	}
+
+	// Calculate the gas required for the keccak256 computation of the input data.
+	gasCost, err := calculateKeccakGas(code)
+	if err != nil {
+		return nil, common.ZeroAddr, 0, err
+	}
+
+	// attempt to grind the address
+	contractAddr, remainingGas, err := evm.attemptGrindContractCreation(caller, nonce, gas, gasCost, code)
+	if err != nil {
+		return nil, common.ZeroAddr, 0, err
+	}
+
+	gas = remainingGas
+
 	return evm.create(caller, &codeAndHash{code: code}, gas, value, contractAddr)
 }
 
+// calculateKeccakGas calculates the gas required for performing a keccak256 hash on the given data.
+// It returns the total gas cost and any error that may occur during the calculation.
+func calculateKeccakGas(data []byte) (int64, error) {
+	// Base gas for keccak256 computation.
+	keccakBaseGas := int64(params.Sha3Gas)
+	// Calculate the number of words (rounded up) in the data for gas calculation.
+	wordCount := (len(data) + 31) / 32 // Round up to the nearest word
+	return keccakBaseGas + int64(wordCount)*int64(params.Sha3WordGas), nil
+}
+
+// attemptContractCreation tries to create a contract address by iterating through possible nonce values.
+// It returns the modified data for contract creation and any error encountered.
+func (evm *EVM) attemptGrindContractCreation(caller ContractRef, nonce uint64, gas uint64, gasCost int64, code []byte) (common.Address, uint64, error) {
+	senderAddress := caller.Address()
+
+	// Iterate through possible nonce values to find a suitable contract address.
+	for i := 0; i < params.MaxAddressGrindAttempts; i++ {
+		// Append the iteration count to the data, altering it for each attempt.
+		codeAndHash := &codeAndHash{code: code}
+
+		// Convert i to a [32]byte array.
+		var salt [32]byte
+		binary.BigEndian.PutUint64(salt[24:], uint64(i)+nonce)
```

Comment:
what is this 24 here?

Reply:
wont we overflow? if nonce and i goes more than 8 bytes?

Reply:
For clarity, I recommend using the uint256 library. Here's how you could do it:

`
i256 := uint256.NewInt(uint64(i))
nonce256 := uint256.NewInt(nonce)
salt := new(uint256.Int).Add(i256, nonce256)
contractAddr := crypto.CreateAddress2(senderAddress, salt.Bytes32(), codeAndHash.Hash().Bytes())
`